From 06ad7bace319da3e065ccef5a14ced14fc962ce5 Mon Sep 17 00:00:00 2001
From: Alexey Bychko <abychko@gmail.com>
Date: Fri, 13 Feb 2026 15:50:43 +0200
Subject: [PATCH 1/2] =?UTF-8?q?[TEST=20INT02]=20Add=20innodb=5Flarge=5Fpag?=
 =?UTF-8?q?e=5Fpopulate=20option=20to=20control=20huge=20page=20pre?=
 =?UTF-8?q?=E2=80=91population?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Introduce a new InnoDB global boolean option innodb_large_page_populate
to control whether huge pages used for the buffer pool are pre-populated at startup.
When enabled (default), behavior remains unchanged and large_page_aligned_alloc
uses MAP_POPULATE and prefaulting. When disabled, the allocator skips MAP_POPULATE
and prefaulting, significantly reducing startup time on large NUMA systems
with huge pages while keeping the existing code path as the default.
---
 storage/innobase/handler/ha_innodb.cc                 | 6 ++++++
 storage/innobase/include/detail/ut/large_page_alloc.h | 7 +++++--
 storage/innobase/srv/srv0srv.cc                       | 1 +
 3 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/storage/innobase/handler/ha_innodb.cc b/storage/innobase/handler/ha_innodb.cc
index a6fd8213be0..ee506d922c6 100644
--- a/storage/innobase/handler/ha_innodb.cc
+++ b/storage/innobase/handler/ha_innodb.cc
@@ -23652,6 +23652,11 @@ static MYSQL_SYSVAR_ULONG(buffer_pool_parallel_init_threads, srv_buffer_pool_par
                           PLUGIN_VAR_RQCMDARG, 
                           "Maximum threads for parallel buffer pool initialization (0=auto)", 
                           nullptr, nullptr, 0, 0, 128, 0);
+
+static MYSQL_SYSVAR_BOOL(large_page_populate, innodb_large_page_populate,
+                          PLUGIN_VAR_NOCMDARG,
+                          "Populate huge pages for InnoDB buffer pool at startup",
+                          nullptr, nullptr, true);
 #endif
 static MYSQL_SYSVAR_ULONG(force_recovery, srv_force_recovery,
                           PLUGIN_VAR_RQCMDARG | PLUGIN_VAR_READONLY,
@@ -24365,6 +24370,7 @@ static MYSQL_SYSVAR_BOOL(encrypt_online_alter_logs,
 static SYS_VAR *innobase_system_variables[] = {
 #ifdef UNIV_LINUX
     MYSQL_SYSVAR(buffer_pool_parallel_init_threads),
+    MYSQL_SYSVAR(large_page_populate),
 #endif
     MYSQL_SYSVAR(api_trx_level),
     MYSQL_SYSVAR(api_bk_commit_interval),
diff --git a/storage/innobase/include/detail/ut/large_page_alloc.h b/storage/innobase/include/detail/ut/large_page_alloc.h
index 4c81c2ca9ed..4b401d05c0a 100644
--- a/storage/innobase/include/detail/ut/large_page_alloc.h
+++ b/storage/innobase/include/detail/ut/large_page_alloc.h
@@ -50,6 +50,7 @@ this program; if not, write to the Free Software Foundation, Inc.,
 #include "storage/innobase/include/detail/ut/pfs.h"
 
 extern const size_t large_page_default_size;
+extern bool innodb_large_page_populate;
 
 namespace ut {
 namespace detail {
@@ -112,7 +113,8 @@ struct Large_page_alloc : public allocator_traits<false> {
   static inline void *alloc(std::size_t size, bool populate) {
     auto total_len = round_to_next_multiple(
         size + page_allocation_metadata::len, large_page_default_size);
-    auto mem = large_page_aligned_alloc(total_len, populate);
+    bool do_populate = populate && innodb_large_page_populate;
+    auto mem = large_page_aligned_alloc(total_len, do_populate);
     if (unlikely(!mem)) return nullptr;
     page_allocation_metadata::datalen(mem, total_len);
     page_allocation_metadata::page_type(mem, Page_type::large_page);
@@ -249,7 +251,8 @@ struct Large_page_alloc_pfs : public allocator_traits<true> {
       page_allocation_metadata::pfs_metadata::pfs_memory_key_t key) {
     auto total_len = round_to_next_multiple(
         size + page_allocation_metadata::len, large_page_default_size);
-    auto mem = large_page_aligned_alloc(total_len, populate);
+    bool do_populate = populate && innodb_large_page_populate;
+    auto mem = large_page_aligned_alloc(total_len, do_populate);    
     if (unlikely(!mem)) return nullptr;
 
 #ifdef HAVE_PSI_MEMORY_INTERFACE
diff --git a/storage/innobase/srv/srv0srv.cc b/storage/innobase/srv/srv0srv.cc
index f74b512af5a..075cc57ed6c 100644
--- a/storage/innobase/srv/srv0srv.cc
+++ b/storage/innobase/srv/srv0srv.cc
@@ -877,6 +877,7 @@ bool srv_buffer_pool_load_at_startup = true;
 #ifdef UNIV_LINUX
 /* The number of init threads */
 ulong srv_buffer_pool_parallel_init_threads = 0;
+bool innodb_large_page_populate = true;
 #endif
 
 /** Slot index in the srv_sys->sys_threads array for the purge thread. */
-- 
2.43.0

